### PR TITLE
feat: allow numbers in component names

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

- Allows for digits in component names
- Refactor to use same validation for snap and component names
- Begin snap and component names errors with a lowercase character

Depends on https://github.com/canonical/craft-parts/pull/790

Fixes [LP#2069783](https://bugs.launchpad.net/snapcraft/+bug/2069783)
(CRAFT-3043)

### CODE REVIEW

**Comment by mr-cal** [comment]
CI is [failing](https://github.com/canonical/snapcraft/actions/runs/10169761191/job/28127405738?pr=4945#step:3:457) because [this reference](https://github.com/canonical/craft-parts/pull/765/files#diff-b6e3778d3bac2384f74b79413bbe8a9cb5f7d35541f2693ea45bc228531f85c8R40) is excluded [here](https://github.com/canonical/snapcraft/blob/feed34ebac8472954d438e2d70beeae0d3ae640e/docs/conf.py#L68)

**Review by mattculler — APPROVED** [review]

**Review by cmatsuoka — APPROVED** [review]

File: tests/unit/models/test_projects.py [hunk 1]
```diff
@@ -2505,26 +2505,21 @@ def test_component_name_valid(
     @pytest.mark.parametrize(
         "name,error",
         [
-            (
-                "snap-",
-                "Component names cannot start with the reserved namespace 'snap-'",
-            ),
-            (
+            pytest.param(
```
Comment (cmatsuoka) [inline tests/unit/models/test_projects.py:2508]: We can test name0123 and 0123name as valid component names above.

**Comment by cmatsuoka** [comment]
The implementation seems good, but we should test valid component names using names containing numbers.

File: tests/unit/models/test_projects.py (hunk 1 above)
Comment (mr-cal) [reply tests/unit/models/test_projects.py:2516]: Added!

**Pull request merged**